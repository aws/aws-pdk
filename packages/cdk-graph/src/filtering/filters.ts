/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { ConstructOrder } from "constructs";
import memoize = require("lodash.memoize"); // eslint-disable-line @typescript-eslint/no-require-imports
import { FilterStrategy, IGraphFilter, IGraphStoreFilter } from "./types";
import { findReferencesOfSubGraph } from "./utils";
import { FlagEnum, Graph, NodeTypeEnum } from "../core";

/**
 * Verify that store is filterable, meaning it allows destructive mutations.
 * @throws Error if store is not filterable
 */
export function verifyFilterable(store: Graph.Store): void {
  if (!store.allowDestructiveMutations) {
    throw new Error(
      "Store must allow destructive mutations to perform filtering; clone the store before applying filters using `store.clone(true)` operation and passing the cloned store to filtering operation."
    );
  }
}

export namespace Filters {
  /**
   * Prune **extraneous** nodes and edges
   * @throws Error if store is not filterable
   * @destructive
   */
  export function pruneExtraneous(): IGraphStoreFilter {
    return (store) => {
      verifyFilterable(store);

      const extraneousNodes = store.root.findAll({
        order: ConstructOrder.POSTORDER,
        predicate: (node) => node.isExtraneous,
      });
      // collapse all extraneous nodes to nearest non-extraneous parent, or prune the node
      for (const extraneousNode of extraneousNodes) {
        const nonExtraneousAncestor = extraneousNode.findAncestor(
          (node) => !node.isExtraneous
        );
        if (nonExtraneousAncestor && !nonExtraneousAncestor.isGraphContainer) {
          extraneousNode.mutateCollapseTo(nonExtraneousAncestor);
        } else {
          extraneousNode.mutateDestroy();
        }
      }

      store.edges.forEach((edge) => {
        if (edge.isExtraneous) {
          edge.mutateDestroy();
        }
      });
    };
  }

  /**
   * Collapses all Cdk Owned containers, which more closely mirrors the application code
   * by removing resources that are automatically created by cdk.
   */
  export function collapseCdkOwnedResources(): IGraphStoreFilter {
    return (store) => {
      store.root
        .findAll({
          order: ConstructOrder.POSTORDER,
          predicate: (node) =>
            Graph.ResourceNode.isResourceNode(node) && node.isCdkOwned,
        })
        .forEach((node) => {
          if (node.isDestroyed) return;
          node.mutateCollapse();
        });
    };
  }

  /**
   * Collapses all Cdk Resource wrappers that wrap directly wrap a CfnResource.
   * Example, s3.Bucket wraps s3.CfnBucket.
   */
  export function collapseCdkWrappers(): IGraphStoreFilter {
    return (store) => {
      const cdkResources = store.root.findAll({
        order: ConstructOrder.POSTORDER,
        predicate: (node) => Graph.ResourceNode.isResourceNode(node),
      }) as Graph.ResourceNode[];
      // collapse all cfnResource wrapped by cdk resource
      for (const cdkResource of cdkResources) {
        if (cdkResource.isWrapper) {
          cdkResource.mutateCollapse();
        } else if (cdkResource.cfnResource) {
          cdkResource.cfnResource.mutateCollapseToParent();
        }
      }
    };
  }

  /**
   * Collapses Custom Resource nodes to a single node.
   */
  export function collapseCustomResources(): IGraphStoreFilter {
    return (store) => {
      store.root
        .findAll({
          predicate: (node) => {
            return node.hasFlag(FlagEnum.CUSTOM_RESOURCE);
          },
        })
        .forEach((customResource) => {
          if (customResource.isDestroyed) return;

          customResource.mutateCollapse();

          if (
            !customResource.hasFlag(FlagEnum.AWS_CUSTOM_RESOURCE) &&
            !customResource.parent?.hasFlag(FlagEnum.AWS_CUSTOM_RESOURCE)
          ) {
            // Try to find resources that are utilized only for the custom resource
            findReferencesOfSubGraph(customResource)
              .filter((_ref, _index, _refs) => {
                // ignore reference that are already children
                if (
                  _ref === customResource ||
                  _ref.isAncestor(customResource)
                ) {
                  return false;
                }
                // ignore refs that either have references or are referenced by nodes outside the set of references
                const _externalByRefs = _ref.referencedBy.filter(
                  (_by) =>
                    _by !== customResource && _by.isAncestor(customResource)
                );
                if (
                  new Set([..._refs, ..._ref.references, ..._externalByRefs])
                    .size !== _refs.length
                ) {
                  return false;
                }
                return true;
              })
              .forEach((_ref) => _ref.mutateMove(customResource));

            customResource.mutateCollapse();
          }
        });
    };
  }

  /**
   * Prune Custom Resource nodes.
   */
  export function pruneCustomResources(): IGraphStoreFilter {
    return (store) => {
      store.root
        .findAll({
          predicate: (node) => {
            return node.hasFlag(FlagEnum.CUSTOM_RESOURCE);
          },
        })
        .forEach((customResource) => {
          if (customResource.isDestroyed) return;

          customResource.mutateDestroy();
        });
    };
  }

  /**
   * Prune empty containers, which are non-resource default nodes without any children.
   *
   * Generally L3 constructs in which all children have already been pruned, which
   * would be useful as containers, but without children are considered extraneous.
   */
  export function pruneEmptyContainers(): IGraphStoreFilter {
    return (store) => {
      store.root
        .findAll({
          predicate: (node) => {
            return node.nodeType === NodeTypeEnum.DEFAULT && node.isLeaf;
          },
        })
        .forEach((node) => {
          if (node.isDestroyed) return;

          node.mutateDestroy();
        });
    };
  }

  /**
   * Collapses extraneous nodes to parent and cdk created nodes on themselves,
   * and prunes extraneous edges.
   *
   * This most closely represents the developers code for the current application
   * and reduces the noise one expects.
   *
   * Invokes:
   * 1.
   * 1. pruneExtraneous()(store);
   * 1. collapseCdkOwnedResources()(store);
   * 1. collapseCdkWrappers()(store);
   * 1. collapseCustomResources()(store);
   * 1. pruneCustomResources()(store);
   * 1. pruneEmptyContainers()(store);
   *
   * @throws Error if store is not filterable
   * @destructive
   */
  export function compact(): IGraphStoreFilter {
    return (store) => {
      verifyFilterable(store);

      pruneExtraneous()(store);
      collapseCdkOwnedResources()(store);
      collapseCdkWrappers()(store);
      collapseCustomResources()(store);
      pruneCustomResources()(store);
      pruneEmptyContainers()(store);
    };
  }

  function _filterNodeType(
    values: (string | RegExp)[],
    exclude: boolean
  ): IGraphStoreFilter {
    const isMatch = memoize((input: string) => {
      if (input == null) {
        return false;
      }

      return (
        values.find((_value) => {
          if (typeof _value === "string") {
            return input === _value;
          }
          return _value.test(input);
        }) != null
      );
    });

    return (store) => {
      for (const node of store.root.findAll({
        order: ConstructOrder.POSTORDER,
      })) {
        if (isMatch(node.nodeType) === exclude) {
          if (node.isLeaf) {
            node.mutateCollapseToParent();
          } else {
            node.mutateUncluster();
          }
        }
      }
    };
  }

  /**
   * Prune all {@link Graph.Node}s *except those matching* specified list.
   *
   * This filter targets all nodes (except root) - {@link IGraphFilter.allNodes}
   * @throws Error if store is not filterable
   * @destructive
   */
  export function includeNodeType(
    nodeTypes: (string | RegExp)[]
  ): IGraphStoreFilter {
    return _filterNodeType(nodeTypes, false);
  }

  /**
   * Prune all {@link Graph.Node}s *matching* specified list.
   *
   * This filter targets all nodes (except root) - {@link IGraphFilter.allNodes}
   * @throws Error if store is not filterable
   * @destructive
   */
  export function excludeNodeType(
    nodeTypes: (string | RegExp)[]
  ): IGraphStoreFilter {
    return _filterNodeType(nodeTypes, true);
  }

  function _filterCfnType(
    values: (string | RegExp)[],
    exclude: boolean
  ): IGraphFilter {
    const isMatch = memoize((input: string) => {
      if (input == null) {
        return false;
      }

      return (
        values.find((_value) => {
          if (typeof _value === "string") {
            return input === _value;
          }
          return _value.test(input);
        }) != null
      );
    });

    return {
      strategy: FilterStrategy.PRUNE,
      node: (node) => {
        // Preserve container structure (stages, stacks, etc.)
        if (node.isCluster || node.isGraphContainer) return true;
        if (Graph.isResourceLike(node)) {
          const match = !!node.cfnType && isMatch(node.cfnType);
          return (match && !exclude) || (!match && exclude);
        }
        // Preserve non *Resource nodes
        return true;
      },
    };
  }

  /**
   * Prune all {@link Graph.ResourceNode} and {@link Graph.CfnResourceNode} nodes
   * *except those matching* specified list of CloudFormation types.
   * @throws Error if store is not filterable
   * @destructive
   */
  export function includeCfnType(cfnTypes: (string | RegExp)[]): IGraphFilter {
    return _filterCfnType(cfnTypes, false);
  }

  /**
   * Prune all {@link Graph.ResourceNode} and {@link Graph.CfnResourceNode} nodes
   * *matching* specified list of CloudFormation types.
   * @throws Error if store is not filterable
   * @destructive
   */
  export function excludeCfnType(cfnTypes: (string | RegExp)[]): IGraphFilter {
    return _filterCfnType(cfnTypes, true);
  }

  /**
   * Remove clusters by hoisting their children to the parent of the cluster
   * and collapsing the cluster itself to its parent.
   * @param clusterTypes
   * @throws Error if store is not filterable
   * @see {@link Graph.Node.mutateUncluster}
   * @destructive
   */
  export function uncluster(clusterTypes?: NodeTypeEnum[]): IGraphStoreFilter {
    // Use set for constant lookup
    const clusterTypesSet = new Set<NodeTypeEnum>(clusterTypes);

    return (store): void => {
      verifyFilterable(store);

      const clusters = store.root.findAll({
        predicate: (node) => {
          if (node.isGraphContainer) return false;
          if (clusterTypesSet.size === 0) return node.isCluster;
          return clusterTypesSet.has(node.nodeType);
        },
      });

      for (const cluster of clusters) {
        cluster.mutateUncluster();
      }
    };
  }
}
