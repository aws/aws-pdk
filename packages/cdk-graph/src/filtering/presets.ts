/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { ConstructOrder } from "constructs";
import { FlagEnum, Graph } from "../core";
import { IGraphFilterPlan } from "./types";

/**
 * Verify that store is filterable, meaning it allows destructive mutations.
 * @throws Error if store is not filterable
 * @internal
 */
export function verifyFilterable(store: Graph.Store): void {
  if (!store.allowDestructiveMutations) {
    throw new Error(
      "Store must allow destructive mutations to perform filtering; clone the store before applying filters using `store.clone(true)` operation and passing the cloned store to filtering operation."
    );
  }
}

/**
 * Changes the root of the store based on filter plan.
 * @throws Error if store is not filterable
 * @internal
 * @destructive
 */
export function rerootFilter(store: Graph.Store, plan: IGraphFilterPlan): void {
  verifyFilterable(store);

  if (plan.root == null) return; // noop

  const planRoot =
    typeof plan.root === "function" ? plan.root(store) : plan.root;

  if (planRoot === store.root) return; // noop

  const hoist = !!plan.hoistRoot;

  const ancestors = planRoot.scopes.slice();
  // remove the actual store.root from ancestors (we can't destroy that)
  const rootAncestor = ancestors.shift();

  if (rootAncestor !== store.root) {
    throw new Error(
      `${planRoot} is not within the store root graph: it has root of ${rootAncestor}`
    );
  }

  if (hoist) {
    // Move plan root as direct child of store root and prune all other ancestors
    planRoot.mutateHoist(store.root);
    // Only need to destroy to first non-root ancestor to prune the ancestral tree
    if (ancestors.length) {
      ancestors[0].mutateDestroy();
    }
    // prune all other root children
    store.root.children.forEach((child) => {
      if (child !== planRoot) {
        child.mutateDestroy();
      }
    });
  } else {
    // keep the plan root in place, but prune non-direct ancestor chain nodes
    // the direct ancestor chain is only the nodes scopes
    ancestors.reverse().forEach((ancestor) => {
      ancestor.siblings.forEach((ancestorSibling) => {
        ancestorSibling.mutateDestroy();
      });
    });

    // prune all planRoot siblings
    planRoot.siblings.forEach((sibling) => {
      sibling.mutateDestroy();
    });
  }
}

/**
 * Performs **compact** filter preset to store.
 * @throws Error if store is not filterable
 * @internal
 * @destructive
 */
export function compactFilterPreset(store: Graph.Store): void {
  verifyFilterable(store);

  const extraneousNodes = store.root.findAll({
    order: ConstructOrder.POSTORDER,
    predicate: (node) => node.isExtraneous,
  });
  // collapse all extraneous nodes to nearest non-extraneous parent, or prune the node
  for (const extraneousNode of extraneousNodes) {
    const nonExtraneousAncestor = extraneousNode.findAncestor(
      (node) => !node.isExtraneous
    );
    if (nonExtraneousAncestor && !nonExtraneousAncestor.isGraphContainer) {
      extraneousNode.mutateCollapseTo(nonExtraneousAncestor);
    } else {
      extraneousNode.mutateDestroy();
    }
  }

  const cdkOwnedContainers = store.root.findAll({
    order: ConstructOrder.POSTORDER,
    predicate: (node) =>
      node.hasFlag(FlagEnum.CDK_OWNED) &&
      !node.parent?.hasFlag(FlagEnum.CDK_OWNED),
  });
  // collapse all cdk owned containers
  // NB: collapses the graph more closely mirror what developer writes, not what is auto created by cdk
  for (const cdkOwnedContainer of cdkOwnedContainers) {
    cdkOwnedContainer.mutateCollapse();
  }

  const cdkResources = store.root.findAll({
    order: ConstructOrder.POSTORDER,
    predicate: (node) => Graph.ResourceNode.isResourceNode(node),
  }) as Graph.ResourceNode[];
  // collapse all cfnResource wrapped by cdk resource
  for (const cdkResource of cdkResources) {
    if (cdkResource.isResourceWrapper) {
      cdkResource.mutateCollapse();
    } else if (cdkResource.cfnResource) {
      cdkResource.cfnResource.mutateCollapseToParent();
    }
  }

  store.edges.forEach((edge) => {
    if (edge.isExtraneous) {
      edge.mutateDestroy();
    }
  });
}
