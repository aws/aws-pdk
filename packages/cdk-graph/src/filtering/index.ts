/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { ConstructOrder } from "constructs";
import { verifyFilterable } from "./filters";
import * as presets from "./presets";
import { FilterPreset, FilterStrategy, IGraphFilterPlan } from "./types";
import { Graph } from "../core";

export * from "./types";
export * from "./filters";

/**
 * Perform graph filter plan on store.
 *
 * This operation is performed *in-place* and is destructive, therefore can only be applied
 * on stores that allow destructive mutations.
 * @param {Graph.Store} store - The store to perform filter plan against
 * @param {IGraphFilterPlan} plan - The filter plan to apply
 * @throws Error is store does not allow destructive mutations
 * @destructive
 */
export function performGraphFilterPlan(
  store: Graph.Store,
  plan: IGraphFilterPlan
): void {
  verifyFilterable(store);

  if (plan.focus) {
    presets.focusFilter(store, plan);
  }

  if (plan.preset) {
    switch (plan.preset) {
      case FilterPreset.NON_EXTRANEOUS: {
        presets.nonExtraneousFilterPreset(store);
        break;
      }
      case FilterPreset.COMPACT: {
        presets.compactFilterPreset(store);
        break;
      }
    }
  }

  if (plan.filters) {
    for (const filter of plan.filters) {
      if (typeof filter === "function") {
        // IGraphStoreFilter
        filter(store);
      } else {
        // IGraphFilter
        const inverse = filter.inverse === true;
        const allNodes =
          filter.allNodes != null ? filter.allNodes : plan.allNodes === true;

        const nodes: Graph.Node[] = store.root.findAll({
          order: plan.order || ConstructOrder.PREORDER,
          predicate: (node) => {
            if (allNodes) return true;
            // by default only return Resources and CfnResources
            return Graph.isResourceLike(node);
          },
        });

        if (filter.node) {
          for (const node of nodes) {
            if (node.isDestroyed) continue;

            const match = filter.node(node);

            if ((match && inverse) || (!match && !inverse)) {
              switch (filter.strategy || FilterStrategy.PRUNE) {
                case FilterStrategy.PRUNE: {
                  node.mutateDestroy();
                  break;
                }
                case FilterStrategy.COLLAPSE: {
                  node.mutateCollapse();
                  break;
                }
                case FilterStrategy.COLLAPSE_TO_PARENT: {
                  node.mutateCollapseToParent();
                  break;
                }
              }
            }
          }
        }

        if (filter.edge) {
          const edges: Graph.Edge[] = store.edges;

          for (const edge of edges) {
            if (edge.isDestroyed) continue;

            const match = filter.edge(edge);

            if ((match && inverse) || (!match && !inverse)) {
              edge.mutateDestroy();
            }
          }
        }
      }
    }
  }
}
