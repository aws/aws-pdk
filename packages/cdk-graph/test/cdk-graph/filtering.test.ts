/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { CfnRole } from "aws-cdk-lib/aws-iam";
import { CfnFunction } from "aws-cdk-lib/aws-lambda";
import {
  CdkGraph,
  FilterPreset,
  getConstructUUID,
  Graph,
  NodeTypeEnum,
  performGraphFilterPlan,
} from "../../src";
import { Filters } from "../../src/filtering/filters";
import { MultiFixtureApp } from "../__fixtures__/apps";
import * as testUtils from "./test-utils";

const makeCdkOutdir = async (name: string) =>
  testUtils.makeCdkOutDir("filtering", name);

describe("cdk-graph/filtering", () => {
  let outdir: string;
  let app: MultiFixtureApp;
  let graph: CdkGraph;
  let originStore: Graph.Store;

  beforeAll(async () => {
    outdir = await makeCdkOutdir("filter");

    app = new MultiFixtureApp({ outdir });
    graph = new CdkGraph(app);
    app.synth();
    originStore = graph.graphContext!.store;
  });

  describe("preset", () => {
    describe("compact", () => {
      let store: Graph.Store;
      beforeAll(() => {
        store = originStore.clone();
      });

      it("should have compactable nodes before testing", () => {
        // sanity check to ensure we are testing the preset
        // extraneous
        expect(
          store.root.findAll({ predicate: (node) => node.isExtraneous }).length
        ).toBeGreaterThan(0);
        expect(
          store.root.findAllLinks({ predicate: (edge) => edge.isExtraneous })
            .length
        ).toBeGreaterThan(0);
        // cfn resources
        expect(
          store.root.findAll({
            predicate: (node) => Graph.CfnResourceNode.isCfnResourceNode(node),
          }).length
        ).toBeGreaterThan(0);
      });

      it("should perform compact filter without errors", () => {
        expect(() =>
          performGraphFilterPlan(store, {
            preset: FilterPreset.COMPACT,
          })
        ).not.toThrow();
      });

      it("should prune all extraneous node", () => {
        expect(
          store.root.findAll({ predicate: (node) => node.isExtraneous }).length
        ).toBe(0);
      });

      it("should prune all extraneous edges", () => {
        expect(
          store.root.findAllLinks({ predicate: (edge) => edge.isExtraneous })
            .length
        ).toBe(0);
      });

      it("should collapse all CfnResourceNodes to parent", () => {
        expect(
          store.root.findAll({
            predicate: (node) => Graph.CfnResourceNode.isCfnResourceNode(node),
          }).length
        ).toBe(0);
      });
    });

    describe("focus/hoist", () => {
      let store: Graph.Store;
      let focusedNode: Graph.Node;

      beforeAll(() => {
        store = originStore.clone();
        focusedNode = store.getNode(getConstructUUID(app.stack.lambda));
      });

      it("should perform focus hoist without error", () => {
        expect(() =>
          performGraphFilterPlan(store, {
            focus: focusedNode,
          })
        ).not.toThrow();
      });

      it("should hoist plan root to store root and remove other children of root", () => {
        expect(store.root.children.length).toBe(1);
        expect(store.root.children[0].uuid).toBe(focusedNode.uuid);
      });

      it("should only have plan root nodes in the store", () => {
        expect(store.counts.nodes).toBe(focusedNode.findAll().length);
      });

      it("should only have plan root edges in the store", () => {
        expect(store.counts.edges).toBe(focusedNode.findAllLinks().length);
      });
    });

    describe("focus/no-hoist", () => {
      let store: Graph.Store;
      let focusedNode: Graph.Node;

      beforeAll(() => {
        store = originStore.clone();
        focusedNode = store.getNode(getConstructUUID(app.stack.lambda));
      });
      it("should perform focus hoist without error", () => {
        expect(() =>
          performGraphFilterPlan(store, {
            focus: {
              node: focusedNode,
              noHoist: true,
            },
          })
        ).not.toThrow();
      });

      it("should remove all non-ancestral nodes from store", () => {
        expect(
          store.root.findAll({
            predicate: (node) => {
              if (node === focusedNode) return false;
              if (focusedNode.isAncestor(node)) return false;
              if (node.isAncestor(focusedNode)) return false;
              return true;
            },
          }).length
        ).toBe(0);
      });

      it("should only have plan root nodes and direct ancestors in the store", () => {
        // NB: plus 1 for root
        expect(store.counts.nodes + 1).toBe(
          focusedNode.findAll().length + focusedNode.scopes.length
        );
      });
    });
  });

  describe("filter", () => {
    describe("custom", () => {
      let store: Graph.Store;

      beforeAll(() => {
        store = originStore.clone();
      });

      const filterTypes = [
        CfnFunction.CFN_RESOURCE_TYPE_NAME,
        CfnRole.CFN_RESOURCE_TYPE_NAME,
      ];

      it("should perform filter without error", () => {
        expect(() =>
          performGraphFilterPlan(store, {
            filters: [
              {
                node: (node) =>
                  !!node.cfnType && filterTypes.includes(node.cfnType),
                edge: (edge) => Graph.Reference.isReference(edge),
              },
            ],
          })
        ).not.toThrow();
      });

      it("should only have filter types nodes", () => {
        expect(
          store.root.findAll({
            predicate: (node) =>
              !!node.cfnType && !filterTypes.includes(node.cfnType),
          }).length
        ).toBe(0);

        expect(store.counts.cfnResources).toMatchSnapshot();
      });

      it("should only have filter type edge", () => {
        expect(
          store.root.findAllLinks({
            predicate: (edge) => !Graph.Reference.isReference(edge),
          }).length
        ).toBe(0);
        expect(store.counts.edgeTypes).toMatchSnapshot();
      });
    });

    describe("nodeType", () => {
      describe("include", () => {
        let store: Graph.Store;

        beforeAll(() => {
          store = originStore.clone();
        });

        const include = [NodeTypeEnum.CFN_RESOURCE, NodeTypeEnum.STACK];

        it("should perform filter without error", () => {
          expect(() =>
            performGraphFilterPlan(store, {
              filters: [Filters.includeNodeType(include)],
            })
          ).not.toThrow();
        });

        it("should only have included nodes", () => {
          expect(
            store.root.findAll({
              predicate: (node) =>
                !node.isGraphContainer && !include.includes(node.nodeType),
            }).length
          ).toBe(0);

          expect(store.counts.cfnResources).toMatchSnapshot();
        });
      });

      describe("excludeNodeType", () => {
        let store: Graph.Store;

        beforeAll(() => {
          store = originStore.clone();
        });

        const excluded = [NodeTypeEnum.OUTPUT, NodeTypeEnum.PARAMETER];

        it("should perform filter without error", () => {
          expect(() =>
            performGraphFilterPlan(store, {
              filters: [Filters.excludeNodeType(excluded)],
            })
          ).not.toThrow();
        });

        it("should not have excluded nodes", () => {
          expect(
            store.root.findAll({
              predicate: (node) => excluded.includes(node.nodeType),
            }).length
          ).toBe(0);

          expect(store.counts.cfnResources).toMatchSnapshot();
        });
      });
    });

    describe("cfnType", () => {
      describe("includeCfnType", () => {
        let store: Graph.Store;

        beforeAll(() => {
          store = originStore.clone();
        });

        const filterTypes = [
          CfnFunction.CFN_RESOURCE_TYPE_NAME,
          CfnRole.CFN_RESOURCE_TYPE_NAME,
        ];

        it("should perform filter without error", () => {
          expect(() =>
            performGraphFilterPlan(store, {
              filters: [Filters.includeCfnType(filterTypes)],
            })
          ).not.toThrow();
        });

        it("should only have included types", () => {
          expect(
            store.root.findAll({
              predicate: (node) =>
                !!node.cfnType && !filterTypes.includes(node.cfnType),
            }).length
          ).toBe(0);

          expect(store.counts.cfnResources).toMatchSnapshot();
        });
      });

      describe("excludeCfnType", () => {
        let store: Graph.Store;

        beforeAll(() => {
          store = originStore.clone();
        });

        const filterTypes = [
          CfnFunction.CFN_RESOURCE_TYPE_NAME,
          CfnRole.CFN_RESOURCE_TYPE_NAME,
        ];

        it("should perform filter without error", () => {
          expect(() =>
            performGraphFilterPlan(store, {
              filters: [Filters.excludeCfnType(filterTypes)],
            })
          ).not.toThrow();
        });

        it("should not have excluded types", () => {
          expect(
            store.root.findAll({
              predicate: (node) =>
                !!node.cfnType && filterTypes.includes(node.cfnType),
            }).length
          ).toBe(0);

          expect(store.counts.cfnResources).toMatchSnapshot();
        });
      });

      describe("excludeCfnType", () => {
        let store: Graph.Store;

        beforeAll(() => {
          store = originStore.clone();
        });

        const filterTypes = [
          CfnFunction.CFN_RESOURCE_TYPE_NAME,
          CfnRole.CFN_RESOURCE_TYPE_NAME,
        ];

        it("should perform filter without error", () => {
          expect(() =>
            performGraphFilterPlan(store, {
              filters: [Filters.excludeCfnType(filterTypes)],
            })
          ).not.toThrow();
        });

        it("should not have filter types nodes", () => {
          expect(
            store.root.findAll({
              predicate: (node) =>
                !!node.cfnType && filterTypes.includes(node.cfnType),
            }).length
          ).toBe(0);

          expect(store.counts.cfnResources).toMatchSnapshot();
        });
      });
    });

    describe("uncluster", () => {
      let store: Graph.Store;

      beforeAll(() => {
        store = originStore.clone();
      });

      const clusterType = NodeTypeEnum.NESTED_STACK;

      it("should initially contain cluster of type to filter", () => {
        expect(store.counts.nodeTypes[clusterType]).toBeGreaterThan(0);
      });

      it("should perform filter without error", () => {
        expect(() =>
          performGraphFilterPlan(store, {
            filters: [Filters.uncluster([clusterType])],
          })
        ).not.toThrow();
      });

      it("should not have nodes of filtered cluster type in store", () => {
        expect(store.counts.nodeTypes[clusterType]).toBe(0);
      });
    });
  });
});
