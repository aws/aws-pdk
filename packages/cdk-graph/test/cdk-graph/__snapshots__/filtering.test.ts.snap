// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-graph/filtering filter cfnType excludeCfnType should not have excluded types 1`] = `
Object {
  "AWS::EC2::EIP": 1,
  "AWS::EC2::Instance": 2,
  "AWS::EC2::InternetGateway": 1,
  "AWS::EC2::NatGateway": 1,
  "AWS::EC2::Route": 6,
  "AWS::EC2::RouteTable": 6,
  "AWS::EC2::SecurityGroup": 3,
  "AWS::EC2::Subnet": 6,
  "AWS::EC2::SubnetRouteTableAssociation": 6,
  "AWS::EC2::VPC": 1,
  "AWS::EC2::VPCGatewayAttachment": 1,
  "AWS::IAM::InstanceProfile": 2,
  "AWS::IAM::Role": 0,
  "AWS::Lambda::Alias": 1,
  "AWS::Lambda::Function": 0,
  "AWS::RDS::DBInstance": 1,
  "AWS::RDS::DBSubnetGroup": 1,
  "AWS::S3::Bucket": 2,
  "AWS::SecretsManager::Secret": 1,
  "AWS::SecretsManager::SecretTargetAttachment": 1,
}
`;

exports[`cdk-graph/filtering filter cfnType excludeCfnType should not have filter types nodes 1`] = `
Object {
  "AWS::EC2::EIP": 1,
  "AWS::EC2::Instance": 2,
  "AWS::EC2::InternetGateway": 1,
  "AWS::EC2::NatGateway": 1,
  "AWS::EC2::Route": 6,
  "AWS::EC2::RouteTable": 6,
  "AWS::EC2::SecurityGroup": 3,
  "AWS::EC2::Subnet": 6,
  "AWS::EC2::SubnetRouteTableAssociation": 6,
  "AWS::EC2::VPC": 1,
  "AWS::EC2::VPCGatewayAttachment": 1,
  "AWS::IAM::InstanceProfile": 2,
  "AWS::IAM::Role": 0,
  "AWS::Lambda::Alias": 1,
  "AWS::Lambda::Function": 0,
  "AWS::RDS::DBInstance": 1,
  "AWS::RDS::DBSubnetGroup": 1,
  "AWS::S3::Bucket": 2,
  "AWS::SecretsManager::Secret": 1,
  "AWS::SecretsManager::SecretTargetAttachment": 1,
}
`;

exports[`cdk-graph/filtering filter cfnType includeCfnType should only have included types 1`] = `
Object {
  "AWS::EC2::EIP": 0,
  "AWS::EC2::Instance": 0,
  "AWS::EC2::InternetGateway": 0,
  "AWS::EC2::NatGateway": 0,
  "AWS::EC2::Route": 0,
  "AWS::EC2::RouteTable": 0,
  "AWS::EC2::SecurityGroup": 0,
  "AWS::EC2::Subnet": 0,
  "AWS::EC2::SubnetRouteTableAssociation": 0,
  "AWS::EC2::VPC": 0,
  "AWS::EC2::VPCGatewayAttachment": 0,
  "AWS::IAM::InstanceProfile": 0,
  "AWS::IAM::Role": 5,
  "AWS::Lambda::Alias": 0,
  "AWS::Lambda::Function": 3,
  "AWS::RDS::DBInstance": 0,
  "AWS::RDS::DBSubnetGroup": 0,
  "AWS::S3::Bucket": 0,
  "AWS::SecretsManager::Secret": 0,
  "AWS::SecretsManager::SecretTargetAttachment": 0,
}
`;

exports[`cdk-graph/filtering filter custom should only have filter type edge 1`] = `
Object {
  "DEPENDENCY": 0,
  "REFERENCE": 15,
}
`;

exports[`cdk-graph/filtering filter custom should only have filter types nodes 1`] = `
Object {
  "AWS::EC2::EIP": 0,
  "AWS::EC2::Instance": 0,
  "AWS::EC2::InternetGateway": 0,
  "AWS::EC2::NatGateway": 0,
  "AWS::EC2::Route": 0,
  "AWS::EC2::RouteTable": 0,
  "AWS::EC2::SecurityGroup": 0,
  "AWS::EC2::Subnet": 0,
  "AWS::EC2::SubnetRouteTableAssociation": 0,
  "AWS::EC2::VPC": 0,
  "AWS::EC2::VPCGatewayAttachment": 0,
  "AWS::IAM::InstanceProfile": 0,
  "AWS::IAM::Role": 5,
  "AWS::Lambda::Alias": 0,
  "AWS::Lambda::Function": 3,
  "AWS::RDS::DBInstance": 0,
  "AWS::RDS::DBSubnetGroup": 0,
  "AWS::S3::Bucket": 0,
  "AWS::SecretsManager::Secret": 0,
  "AWS::SecretsManager::SecretTargetAttachment": 0,
}
`;

exports[`cdk-graph/filtering filter nodeType excludeNodeType should not have excluded nodes 1`] = `
Object {
  "AWS::EC2::EIP": 1,
  "AWS::EC2::Instance": 2,
  "AWS::EC2::InternetGateway": 1,
  "AWS::EC2::NatGateway": 1,
  "AWS::EC2::Route": 6,
  "AWS::EC2::RouteTable": 6,
  "AWS::EC2::SecurityGroup": 3,
  "AWS::EC2::Subnet": 6,
  "AWS::EC2::SubnetRouteTableAssociation": 6,
  "AWS::EC2::VPC": 1,
  "AWS::EC2::VPCGatewayAttachment": 1,
  "AWS::IAM::InstanceProfile": 2,
  "AWS::IAM::Role": 6,
  "AWS::Lambda::Alias": 1,
  "AWS::Lambda::Function": 3,
  "AWS::RDS::DBInstance": 1,
  "AWS::RDS::DBSubnetGroup": 1,
  "AWS::S3::Bucket": 2,
  "AWS::SecretsManager::Secret": 1,
  "AWS::SecretsManager::SecretTargetAttachment": 1,
}
`;

exports[`cdk-graph/filtering filter nodeType include should only have included nodes 1`] = `
Object {
  "AWS::EC2::EIP": 1,
  "AWS::EC2::Instance": 2,
  "AWS::EC2::InternetGateway": 1,
  "AWS::EC2::NatGateway": 1,
  "AWS::EC2::Route": 6,
  "AWS::EC2::RouteTable": 6,
  "AWS::EC2::SecurityGroup": 3,
  "AWS::EC2::Subnet": 6,
  "AWS::EC2::SubnetRouteTableAssociation": 6,
  "AWS::EC2::VPC": 1,
  "AWS::EC2::VPCGatewayAttachment": 1,
  "AWS::IAM::InstanceProfile": 2,
  "AWS::IAM::Role": 6,
  "AWS::Lambda::Alias": 1,
  "AWS::Lambda::Function": 3,
  "AWS::RDS::DBInstance": 1,
  "AWS::RDS::DBSubnetGroup": 1,
  "AWS::S3::Bucket": 2,
  "AWS::SecretsManager::Secret": 1,
  "AWS::SecretsManager::SecretTargetAttachment": 1,
}
`;
